src/Functions/map.cpp

```diff
@@ -144,6 +146,97 @@ class FunctionMap : public IFunction
+    static constexpr auto name = "mapFromArrays";
+
+    static FunctionPtr create(ContextPtr)
+    {
+        return std::make_shared<FunctionMapFromArrays>();
+    }
+
+    String getName() const override
+    {
+        return name;
+    }
+
+    size_t getNumberOfArguments() const override
+    {
+        return 2;
+    }
+
+    bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo & /*arguments*/) const override { return false; }
+    bool useDefaultImplementationForNulls() const override { return false; }
+    bool useDefaultImplementationForConstants() const override { return true; }
+
+    DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override
+    {
+        if (arguments.size() != 2)
+            throw Exception(ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH,
+                "Function {} requires 2 arguments, but {} given", getName(), arguments.size());
+
+        const auto * keys_type = checkAndGetDataType<DataTypeArray>(arguments[0].get());
+        if (!keys_type)
+            throw Exception{"First argument for function " + getName() + " must be a Array",
+                            ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT};
+
+        const auto * values_type = checkAndGetDataType<DataTypeArray>(arguments[1].get());
+        if (!values_type)
+            throw Exception{"Second argument for function " + getName() + " must be a Array",
+                            ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT};
+
+        DataTypes tmp;
+        const auto & key_type = keys_type->getNestedType();
+        const auto & value_type = values_type->getNestedType();
+        tmp.emplace_back(key_type);
+        tmp.emplace_back(value_type);
+        return std::make_shared<DataTypeMap>(tmp);
+    }
+
+    ColumnPtr executeImpl(
+        const ColumnsWithTypeAndName & arguments, const DataTypePtr & /* result_type */, size_t /* input_rows_count */) const override
+    {
+        bool is_keys_const = isColumnConst(*arguments[0].column);
+        const ColumnArray * col_keys;
+        if (is_keys_const)
+        {
+            auto tmp = arguments[0].column->convertToFullColumnIfConst();
+            col_keys = checkAndGetColumn<ColumnArray>(tmp.get());
+        }
+        else
+        {
+            col_keys = checkAndGetColumn<ColumnArray>(arguments[0].column.get());
+        }
+
+        bool is_values_const = isColumnConst(*arguments[1].column);
+        const ColumnArray * col_values;
+        if (is_values_const)
+        {
+            auto tmp = arguments[1].column->convertToFullColumnIfConst();
+            col_values = checkAndGetColumn<ColumnArray>(tmp.get());
+        }
+        else
+        {
+            col_values = checkAndGetColumn<ColumnArray>(arguments[1].column.get());
+        }
+
+        if (!col_keys || !col_values)
+            throw Exception("Arguments of function " + getName() + " must be array.", ErrorCodes::ILLEGAL_COLUMN);
+
+        if (!col_keys->hasEqualOffsets(*col_values))
```

Comment:
Why do we check equal offsets, but not just sizes?

Reply:
Because in each row of `col_keys` and `col_values`,  their values whose type is Array must have equal size, otherwise we can't create a map. For example: `select mapFromArrays([1,2,3], ['a', 'b'])`